x-pack/legacy/plugins/lens/public/editor_frame_plugin/plugin.tsx

```diff
@@ -41,18 +43,35 @@ export class EditorFramePlugin {
   private readonly datasources: Record<string, Datasource> = {};
   private readonly visualizations: Record<string, Visualization> = {};
 
+  private embeddableFactory: EmbeddableFactory | null = null;
+
   public setup(_core: CoreSetup | null, plugins: EditorFrameSetupPlugins): EditorFrameSetup {
-    plugins.interpreter.functionsRegistry.register(() => mergeTables);
-
-    plugins.embeddables.registerEmbeddableFactory(
-      'lens',
-      new EmbeddableFactory(
-        plugins.chrome,
-        plugins.data.expressions.ExpressionRenderer,
-        plugins.data.indexPatterns.indexPatterns
-      )
+    plugins.data.expressions.registerFunction(() => mergeTables);
+
+    this.embeddableFactory = new EmbeddableFactory(
+      plugins.chrome,
+      plugins.data.indexPatterns.indexPatterns
     );
 
+    plugins.embeddables.registerEmbeddableFactory('lens', this.embeddableFactory);
+
+    return {
+      registerDatasource: (name, datasource) => {
+        this.datasources[name] = datasource as Datasource<unknown, unknown>;
+      },
+      registerVisualization: visualization => {
+        this.visualizations[visualization.id] = visualization as Visualization<unknown, unknown>;
+      },
+    };
+  }
+
+  public start(_core: CoreStart | null, plugins: EditorFrameStartPlugins): EditorFrameStart {
+    if (this.embeddableFactory === null) {
```

Comment:
Maybe it's not possible due to the way our lifecycle methods work, but could `plugins.data.expressions.ExpressionRenderer` be passed into the embeddable factory's constructor?